### ISSUE TITLE
Potentially incorrect hypothesis in "How Atoms differ from Elegant Objects"

### ISSUE THREAD

**Issue by victornoel** [project-avral/oo-atom#127]
I was reading the ["How Atoms differ from Elegant Objects"](https://github.com/project-avral/oo-atom/wiki/How-Atoms-differ-from-Elegant-Objects) page and found that the argument about "Inheritance vs subtyping" was not founded.

The main problem underlined is that with final-only classes, you can't add new constructor in other classes.
This is IMHO incorrect: this is the whole purpose of using an abstract Wrapper class (see http://www.yegor256.com/2017/01/31/decorating-envelopes.html for examples).

Basically for each of your interface, you declare a `Wrap` class which role is only to delegate to another instance of this very interface, and then if you need a new constructor for your class, you simply extend the `Wrap` class and wrap inside an instance of the original implementation with your own parameters passed to the primary constructor.

This is verbose because of the way Java works, but it does seem to solve the problem underlined in the wiki page.

What do you think?

**Comment by skapral** [comment]
@victornoel Year, seems like I missed that post initially. When I wrote the page you mentioned, I referenced mainly the Elegant Objects book, which was quite straight about it. In Atom concept I have my own analogue for decorating envelopes - I call them Atom aliases (see the [spec](docs/ATOM_SPECIFICATION.md). Rereading the post now, it looks like I reinvented the same thing - probably I should mention it somewhere.

However, while experimenting on elegant objects, I came to a conclusion which works for me and contradicts with EO - there is almost always no sense in making "elegant" class final. Final classes are harder to extend. By extendability here I mean "Open-Closed" principle from SOLID, and not implementation inheritance. And - yes, I know [Yegor's](http://www.yegor256.com/2017/03/28/solid.html) opinion on SOLID, but totally disagree with him on OCP part - OCP per-se tells nothing about implementation inheritance.

BTW, abstract Atoms I restricted too. For me, it seems like when developer uses an abstract class, he tries to reuse some functinality in several other classes (modules), which will likely mean breaking Single Responcibility principle for them. In my practice, if things were decomposed correctly, there is no need in abstract classes ever.

So, instead of making all elegant classes to be either final or abstract, like "Elegant Objects" book tells in section 4.3, I made a totally opposite rule - I restrict Atoms to be final, and restrict them to be abstract. That point I tried to emphasize in the framework and the Wiki page.
